Added Multipart Upload Specifications (#19)
ghga-de/epic-docs: 4-star-nosed_mole/multipart_upload_specifications.md

# Up-/Download Client & Production POC (Star-Nosed Mole) - Multipart Uploads

This is an addition to the Star-Nosed mole epic to adress challenges in multipart uploads

## Scope:
A scope definition can be found here: https://wiki.verbis.dkfz.de/x/fAFzCQ

## User Journeys

The addition to this epic covers the following user journeys:

### Multipart Data Upload:

> ![Data Upload](./images/multipart_data_upload.png)

Reviewer: ```suggestion
![Data Upload](./images/multipart_data_upload.jpg)
```

> A Data Submitter specifies the file ID, the URL to the Upload Contoller API, and the file path on the local file system using the CLI interface of the client (1.0). Internally, the CLI client translates the user-defined data into a request to the Upload Contoller API to obtain an S3 upload id (1.1). The client sends a request to the Upload Controller for each individual part of the uploaded file and recieves a presigned post for each individual part (1.2). The client reads multiple parts from the source file and uploads them as a stream using the pre-signed post (1.3). Once the upload has been completed the client sends a confirmation to the Upload Controller API (1.4).

Reviewer: ```suggestion
A Data Submitter specifies the file ID, the URL to the Upload Contoller API, and the file path on the local file system using the CLI interface of the client (1.0). Internally, the CLI client translates the user-defined data into a request to the Upload Contoller API to obtain an upload id (1.1). The client sends a request to the Upload Controller for each individual part of the uploaded file and recieves a presigned post for each individual part (1.2). The client reads multiple parts from the source file and uploads them as a stream using the pre-signed post (1.3). Once the upload has been completed the client sends a confirmation to the Upload Controller API thereby specifying the eTag for each file part (1.4).
```

We can just stay generic and say there is an upload ID, we might want to only keep the actual ID we get from S3 private in our database and expose another ID to the public.

## CLI:

The CLI commands remain the same.

## UCS:

The following Updates are performed in the Upload Controller Service:

### Database:

A new table "multipart_uploads" is added, containing the following columns:
```

> - s3_upload_id (Primary key)

Reviewer: Maybe rather:
```suggestion
- upload_id (Primary Key, UUID)
- s3_upload_id (unique)
```

- file_id (not unique)
- upload_status
```

> file_state is an Enum and can have the following values:

Reviewer: ```suggestion
Thereby, the upload_status is an Enum and can have the following values:
```

```
- pending
- canceled
- failed
- uploaded
- accepted
- rejected
```

> The column "upload_status" from then table "files" is removed.

Reviewer: ```suggestion
Please note, not more that one upload per file may have a state that is set to `pending` , `uploaded`, or `accepted`. Morover, within the list of states from uploads corresponding to one file, these `pending` , `uploaded`, and `accepted` are mutually exclusive.

This also means that once an upload attempt corresponding to a file is set to `accepted`, no new uploads can be created for that file. In a future epic, we will implement a mechanism that allows controlled re-upload of a file by explicitly requesting to depreciate an old upload. We might also consider an `is_open` flag that is specified per file to control whether new upload attempts are currently allowed for that file.

The column "upload_status" from the table "files" is removed.
```

> A column "s3_upload_id" (unique) is added to the table "files".

Reviewer: I think this is not needed, because the "current" upload attempt can be reconstructed from the uploads table.

### Endpoints:

Replace:
```
GET: /presigned_post/{file_id}
    gets presigned_post from S3
    sets UploadState to Pending
returns presigned_post
```

with:
```
POST /files/{file_id}/multipart_uploads
    if there is already a multipart_upload with file_id and upload_status == pending, updated or accepted
        returns error
    else:
        gets s3_upload_id from S3
        creates new row in multipart_uploads with file_id, s3_upload_id and upload_status=pending
        put s3_upload_id into database

> returns s3_upload_id and the part size in bytes for this download (default: 16 MiB)

Reviewer: ```suggestion
returns upload_id and the part size in bytes for this download (default: 16 MiB)
```

Reviewer: We don't know the exact size of the file before we upload right? Could we thus give instructions on the part size that shall be used?

Author: I assumed, we get those information from the metadata service

```

Add:
```
GET /files/{file_id}/multipart_uploads?status=pending

> returns the s3_upload_id in the table multipart_uploads where file_id=file_id and upload_status=pending (There should never be more than one.) or None

Reviewer: ```suggestion
returns the upload_id in the table multipart_uploads where file_id=file_id and upload_status=pending (There should never be more than one.) or None
```

Author: The CLI needs the s3_upload_id to upload to s3.

```

Add:
```

> POST /multipart_uploads/{upload_id}/part/{part_no}/presigned_post

Reviewer: Should also be plural for REST-compliance.
We might also consider to rename it to simple `.../presigned_urls` or `.../urls`.

```suggestion
POST /multipart_uploads/{upload_id}/part/{part_no}/presigned_posts
```

Author: But what we recieve are in fact not urls, but an url with additional information we have to send in the post body, so url doesn't fit here.

    creates part-specific presigned post
returns presigned post
```

Replace:
```
patch: /confirm_upload/{file_id}
    body: state: UploadState (Uploaded)
    sets upload_status to uploaded
returns: 204
```

with:
```
PATCH /multipart_uploads/{upload_id}/
    body:       status==uploaded/cancelled

>                 part_no to eTag mapping (only if status==uploaded)

Reviewer: Maybe we can also store that mapping in the database?

Author: Why would we need to?
